metals/src/main/scala/scala/meta/internal/metals/onTypeRangeFormatters/OnTypeRangeFormatter.scala

```diff
@@ -0,0 +1,55 @@
+package scala.meta.internal.metals.onTypeRangeFormatters
+
+import scala.meta.tokens.Tokens
+
+import org.eclipse.lsp4j.FormattingOptions
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import org.eclipse.lsp4j.TextEdit
+
+sealed trait PositionTyped {
+  type StartPosition = meta.Position
+  type EndPosition = meta.Position
+}
+
+trait OnTypeFormatter extends PositionTyped {
+
+  def contribute(
+      onTypeformatterParams: OnTypeFormatterParams,
+      tokensParams: TokensParams
+  ): Option[List[TextEdit]] = None
+
+}
+
+trait RangeFormatter extends PositionTyped {
+  def contribute(
+      rangeFormatterParams: RangeFormatterParams,
+      tokensParams: TokensParams
+  ): Option[List[TextEdit]] = None
+}
+
+case class TokensParams(
+    startPos: meta.Position,
+    endPos: meta.Position,
+    tokens: Option[Tokens]
+)
+
+trait FormatterParams {
+  val sourceText: String
+  val range: Range
+  lazy val splitLines: Array[String] = sourceText.split('\n')
```

Comment:
```suggestion
  lazy val splitLines: Array[String] = sourceText.split(System.lineSeparator)
```

Reply:
This suggestion wasn't correct and leaded to `OnTypeFormattingSuite` failure.
Let's try this one. I checked that `tests.rangeFomatiing.*` + `tess.OnFomattingSuite` pass locally with the following:
```scala
lazy val splitLines: Array[String] = sourceText.split("\\r?\\n")
```

Reply:
What about the `mkString` in file `IndentOnPaste`, should we change it back to `lines.mkString('\n')` to be consistent with the rest of the project?

Reply:
@Giggiux no. We should keep `System.lineSeparator`  there. Otherwise the resulting text will have different sep from expected text on Windows